staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,750 @@
+				klog.V(7).Infof("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, immediate", rd, fs.Name, fs.Spec.DistinguisherMethod, plName)
+				return fs.Name, fs.Spec.DistinguisherMethod, plName, nil
+			}
+			plState.numPending++
+			matchID := &plName
+			klog.V(7).Infof("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID)
+			var startCalls int32
+			startFn := func(ctx context.Context, hashValue uint64) (execute bool, afterExecution func()) {
+				newStarts := atomic.AddInt32(&startCalls, 1)
+				klog.V(7).Infof("For matchID=%p, startFn(ctx, %v) startCalls:=%d", matchID, hashValue, newStarts)
+				if newStarts > 1 {
+					panic(fmt.Sprintf("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p startCalls:=%d", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID, newStarts))
+				}
+				req := func() fq.Request {
+					cfgCtl.lock.Lock()
+					defer cfgCtl.lock.Unlock()
+					plState.numPending--
+					req, idle1 := plState.queues.StartRequest(ctx, hashValue, rd.RequestInfo, rd.User)
+					if idle1 {
+						cfgCtl.maybeReapLocked(plName, plState)
+					}
+					return req
+				}()
+				if req == nil {
+					klog.V(7).Infof("For matchID=%p, startFn(ctx, %v): reject", matchID, hashValue)
+					return false, func() {}
+				}
+				exec, idle2, afterExec := req.Wait()
+				execID := &exec
+				if idle2 {
+					cfgCtl.maybeReap(plName)
+				}
+				klog.V(7).Infof("For matchID=%p, startFn(..) => exec=%v, execID=%p", matchID, exec, execID)
+				var afterCalls int32
+				return exec, func() {
+					newAfter := atomic.AddInt32(&afterCalls, 1)
+					klog.V(7).Infof("For matchID=%p, execID=%p, after() calls:=%d", matchID, execID, newAfter)
+					if newAfter > 1 {
+						panic(fmt.Sprintf("Match(%#+v) => fsName=%q, distMethod=%#+v, plName=%q, matchID=%p, execID=%p, afterCalls:=%d", rd, fs.Name, fs.Spec.DistinguisherMethod, plName, matchID, execID, newAfter))
+					}
+					idle3 := afterExec()
+					if idle3 {
+						cfgCtl.maybeReap(plName)
+					}
+				}
+			}
+			return fs.Name, fs.Spec.DistinguisherMethod, plName, startFn
+		}
+	}
+	var lastFS *fctypesv1a1.FlowSchema
+	if len(cfgCtl.flowSchemas) > 0 {
+		lastFS = cfgCtl.flowSchemas[len(cfgCtl.flowSchemas)-1]
+	}
+	// This can never happen because every configState has a
+	// FlowSchema that matches everything.
+	panic(fmt.Sprintf("No match; rd=%#+v, lastFS=%#+v", rd, fcfmt.Fmt(lastFS)))
+}
+
+// Call this after getting a clue that the given priority level is undesired and idle
+func (cfgCtl *configController) maybeReap(plName string) {
+	cfgCtl.lock.Lock()
+	defer cfgCtl.lock.Unlock()
+	plState := cfgCtl.priorityLevelStates[plName]
+	if plState != nil && (plState.queues == nil || plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle()) {
+		klog.V(3).Infof("Triggered API Priority and Fairness config reload because priority level %s became idle", plName)
+		cfgCtl.configQueue.Add(0)
```

Comment:
There are two reasons why it does not.

(1) Because the caller does not hold the lock, it is not guaranteed that the named priority level still exists in the controller's map.  So that is necessary to check here but not in maybeReapLocked.

(2) Also because the caller does not hold the lock, it is not guaranteed that any clue the caller got about the QueueSet being idle is still valid.  So that also has to be checked here and not in maybeReapLocked.